### PR TITLE
[RFE] Allow multiple groups creation

### LINKED ISSUE #879: [RFE] Allow multiple groups creation
ipauser allows creating/deleting multiple users in one go and ipagroup allows deleting multiple groups in one operation.

In some cases it would speed things up if multiple groups could be created with one operation.

Thanks.

### PR DESCRIPTION
**Pull request by dkarpele**
Adding an option to create multiple groups in one go. 
Adding tests (present/absent)

Signed-off-by: Denis Karpelevich <dkarpele@redhat.com>

Fixes #879

### CODE REVIEW

File: plugins/modules/ipagroup.py [hunk 1]
```diff
@@ -257,6 +340,25 @@ def gen_member_args(user, group, service, externalmember, idoverrideuser):
     return _args
 
 
+def check_parameters(  # pylint: disable=unused-argument
```
Comment (rjeffman) [inline plugins/modules/ipagroup.py:343]: Other than `module`, `state` and `action`, the other parameters are not needed.

File: tests/group/test_groups_absent.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  - name: Create dict with group names
+    set_fact:
+      groups_names: "{{ groups_names | default([]) + [{ 'name': item.name }] }}"
+    loop: "{{ group_list }}"
+
+  - name: groups absent len:{{ group_list | length }}
+    ipagroup:
+      ipaadmin_password: SomeADMINpassword
+      groups: "{{ groups_names }}"
+      state: absent
```
Comment (rjeffman) [inline tests/group/test_groups_absent.yml:24]: Please, clean up file `groups.json` after test completion.

File: tests/group/test_groups_present.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  tasks:
+  - name: Include groups.json
+    include_vars:
+      file: groups.json  # noqa 505
+
+  - name: groups present len:{{ group_list | length }}
+    ipagroup:
+      ipaadmin_password: SomeADMINpassword
+      groups: "{{ group_list }}"
```
Comment (rjeffman) [inline tests/group/test_groups_present.yml:18]: Please, clean up file `groups.json` after test completion.

File: tests/group/test_groups_present_slice.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    include_vars:
+      file: groups.json  # noqa 505
+  - name: Size of groups slice.
+    debug:
+      msg: "{{ group_list | length }}"
+  - name: groups present
+    ipagroup:
+      ipaadmin_password: SomeADMINpassword
+      groups: "{{ group_list[item:item+slice_size] }}"
+    loop: "{{ range(0,group_list | length, slice_size) | list }}"
```
Comment (rjeffman) [inline tests/group/test_groups_present_slice.yml:23]: Please, clean up file `groups.json` after test completion.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@dkarpele thank you for the PR.

It looks good, but please check the suggestions.

**Review by rjeffman — APPROVED** [review]
LGTM.

**Review by varunmylaraiah — APPROVED** [review]
LGTM

**Comment by rjeffman** [comment]
@dkarpele, please squash commits into one using the first commit message.

**Comment by varunmylaraiah** [comment]
@dkarpele 
How can we add a group with a non-posix or external type using JSON? If it is not currently implemented, please implement it."

**Comment by dkarpele** [comment]
> @dkarpele How can we add a group with a non-posix or external type using JSON? If it is not currently implemented, please implement it."

Great catch @varunmylaraiah! This problem is not related to JSON and there is no issue with external type. The failing case is with nonposix. The group `group_nonposix` will be created as a `posixgroup`:
```
 - name: Add nonposix groups
    ipagroup:
      ipaadmin_password: SomeADMINpassword
      groups:
      - name: group_nonposix
        nonposix: yes
```
I'll investigate it.

**Comment by dkarpele** [comment]
Hi @rjeffman and @varunmylaraiah, could you please re-review the PR? I added a patch for bug found by Varun and test for the nonposix/external case.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@dkarpele please check and fix the warnings raised by action "Verify ansible-lint". (It seems they are easy to spot in "Files changed".)

**Comment by dkarpele** [comment]
@rjeffman fixed

**Review by rjeffman — CHANGES_REQUESTED** [review]
@dkarpele also, make the PR a single commit (or at most two, one for the plugin, one for the tests, but one commit is good enough for this one).

**Review by rjeffman — APPROVED** [review]
lgtm

**Comment by varunmylaraiah** [comment]
LGTM.

**Comment by t-woerner** [comment]
Please add examples for multi group use to the readme and also to the EXAMPLES section in the plugin.

**Review by varunmylaraiah — CHANGES_REQUESTED** [review]
This PR has an issue when adding external group with members. 

```
[root@ansible ~]# cat external_with_members.json
{
    "multiplegroups": [
      {
        "name": "066externalgroup",
        "description": "test add external groups with members",
        "external": "yes",
        "user": "gmm_user01",
        "group": ["gmm_group01", "gmm_group02"]
      }
    ]
  }
  
```
  
```
  
  [root@ansible ~]# cat group_module2.yml
---
- name: Playbook to ensure add multiple groups with members using json file.
  hosts: ipaserver
  become: true

  tasks:
  - name: groups add using json file
    include_vars:
      file: external_with_members.json  # noqa missing-import

  - name: Groups
    ipagroup:
      ipaadmin_password: <xxxxxxxxxxx>
      "groups": "{{ multiplegroups }}"
      
```
 
 ```

 PLAYBOOK: group_module2.yml ***********************************************************************************
1 plays in group_module2.yml

PLAY [Playbook to ensure add multiple groups with members using json file.] ***********************************

TASK [Gathering Facts] ****************************************************************************************
task path: /root/group_module2.yml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [groups add using json file] *****************************************************************************
task path: /root/group_module2.yml:7
ok: [master.ipadomain.test] => {"ansible_facts": {"multiplegroups": [{"description": "test add multiple external groups", "external": "yes", "group": ["gmm_group01", "gmm_group02"], "name": "066externalgroup", "user": "gmm_user01"}]}, "ansible_included_var_files": ["/root/external_with_members.json"], "changed": false}

TASK [Groups] *************************************************************************************************
task path: /root/group_module2.yml:11
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "group_add_member: 066externalgroup: Cannot perform external member validation without Samba 4 support installed. Make sure you have installed server-trust-ad sub-package of IPA on the server"}

PLAY RECAP ****************************************************************************************************
master.ipadomain.test      : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by varunmylaraiah** [comment]
or

```
[root@ansible ~]# cat external_group_with_members.yaml
---
- name: Playbook to ensure a external group is present with members
  hosts: ipaserver
  become: true

  tasks:
  # Create group
  - ipagroup:
      ipaadmin_password: <xxxxxxx>
      name: testexternal
      external: yes
  - ipagroup:
      ipaadmin_password:  <xxxxxxx>
      name: testexternal
      action: member
      user: gmm_user01
      
      
 PLAYBOOK: external_group_with_members.yaml ********************************************************************
1 plays in external_group_with_members.yaml

PLAY [Playbook to ensure a external group is present with members] ********************************************

TASK [Gathering Facts] ****************************************************************************************
task path: /root/external_group_with_members.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipagroup] ***********************************************************************************************
task path: /root/external_group_with_members.yaml:8
ok: [master.ipadomain.test] => {"changed": false}

TASK [ipagroup] ***********************************************************************************************
task path: /root/external_group_with_members.yaml:12
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "group_add_member: testexternal: Cannot perform external member validation without Samba 4 support installed. Make sure you have installed server-trust-ad sub-package of IPA on the server"}

PLAY RECAP ****************************************************************************************************
master.ipadomain.test      : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by dkarpele** [comment]
Hi @varunmylaraiah,
1. I think this is not a bug. If you check any test with external members, for example `test_group_external_members.yml` there is a condition: `when: trust_test_is_supported | default(false)`. Please recheck your test with this condtion.
2. The second scenario you mention doesn't have `groups` at all. So, even, if it's a bug - it's not related to current PR.

**Comment by varunmylaraiah** [comment]
Hello @dkarpele @t-woerner,
It appears that there is a problem with the module. After trying the same scenario on the master branch, I observed the same error. However, the group and its members were added. Additionally, I tested it on the CLI but I did not see any error.

[root@master ~]# ipa group-show testexternal
ipa: ERROR: testexternal: group not found

```
---
- name: Playbook to ensure a external group is present with members
  hosts: ipaserver
  become: true

  tasks:
  # Create group
  - ipagroup:
      ipaadmin_password: <xxxxxxx>
      name: testexternal
      external: yes
  - ipagroup:
      ipaadmin_password:  <xxxxxxx>
      name: testexternal
      action: member
      user: gmm_user01

PLAYBOOK: external_group_with_members.yaml ********************************************************************
1 plays in external_group_with_members.yaml

PLAY [Playbook to ensure a external group is present with members] ********************************************

TASK [Gathering Facts] ****************************************************************************************
task path: /root/external_group_with_members.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipagroup] ***********************************************************************************************
task path: /root/external_group_with_members.yaml:8
changed: [master.ipadomain.test] => {"changed": true}

TASK [ipagroup] ***********************************************************************************************
task path: /root/external_group_with_members.yaml:12
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "group_add_member: testexternal: Cannot perform external member validation without Samba 4 support installed. Make sure you have installed server-trust-ad sub-package of IPA on the server"}

PLAY RECAP ****************************************************************************************************
master.ipadomain.test      : ok=2    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0



**Group "testexternal" and its members "gmm_user01" were added**
[root@master ~]# ipa group-show testexternal --all
  dn: cn=testexternal,cn=groups,cn=accounts,dc=ipadomain,dc=test
  Group name: testexternal
  Member users: gmm_user01
  ipauniqueid: 3e42bd6a-b328-11ed-bc82-fa163e349009
  objectclass: top, groupofnames, nestedgroup, ipausergroup, ipaobject,
               ipaexternalgroup

```
**CLI execution:**

```
[root@master ~]# ipa group-show testexternal
ipa: ERROR: testexternal: group not found

[root@master ~]# ipa group-add testexternal --external
--------------------------
Added group "testexternal"
--------------------------
  Group name: testexternal

[root@master ~]# ipa group-add-member testexternal --users=gmm_user01
  Group name: testexternal
  Member users: gmm_user01
-------------------------
Number of members added 1
-------------------------

[root@master ~]# ipa group-show testexternal --all
  dn: cn=testexternal,cn=groups,cn=accounts,dc=ipadomain,dc=test
  Group name: testexternal
  Member users: gmm_user01
  ipauniqueid: 26c22e2a-b32b-11ed-8139-fa163e349009
  objectclass: top, groupofnames, nestedgroup, ipausergroup, ipaobject,
               ipaexternalgroup
```

**Comment by dkarpele** [comment]
Thanks @varunmylaraiah. As you were able to reproduce it in master and without `groups` argument, could you open a separate issue for this problem? To avoid confusion, I don't think we should discuss it here.

**Review by varunmylaraiah — APPROVED** [review]

**Comment by varunmylaraiah** [comment]
Hi @dkarpele 
Please have a look below scenarios.
In the first scenario, I attempted to add three groups using JSON (a non-posix group, a posix group, and another non-posix group). While trying to add the second non-posix group, I encountered the error message: 
`"FAILED! => {"changed": false, "msg": "group_add: 07nonposixgroup: missing attribute "gidNumber" required by object class "posixGroup""}"`

In the second scenario, I successfully added the same non-posix group (07nonposixgroup). 

Please have a look.


First scenario.
```
# cat group_module.yml
---
- name: Playbook to ensure add multiple groups with members using json file.
  hosts: ipaserver
  become: true

  tasks:
  - name: groups add using json file
    include_vars:
      file: group_add_multiple_groups.json  # noqa missing-import

  - name: Groups
    ipagroup:
      ipaadmin_password: <xxxxxxxxxxxx>
      "groups": "{{ multiplegroups }}"
      

# cat group_add_multiple_groups.json
{
  "multiplegroups": [
    {
      "name": "06nonposixgroup",
      "description": "test add nonposix group with members",
      "nonposix": "yes",
      "group": ["gmm_group01", "gmm_group02"]
    },
    {
      "name": "06posixgroup",
      "description": "test add posix group with members",
      "user": ["gmm_user01", "gmm_user02"]
    },
    {
      "name": "07nonposixgroup",
      "description": "test add nonposix group with member",
      "nonposix": "yes",
      "user": ["gmm_user01"],
      "group": ["gmm_group01", "gmm_group02"]
    }
  ]
  
  
PLAYBOOK: group_module.yml **********************************************************************************************************************************************************************************************************
1 plays in group_module.yml

PLAY [Playbook to ensure add multiple groups with members using json file.] *********************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************************************************************
task path: /root/group_module.yml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [groups add using json file] ***************************************************************************************************************************************************************************************************
task path: /root/group_module.yml:7
ok: [master.ipadomain.test] => {"ansible_facts": {"multiplegroups": [{"description": "test add nonposix group with members", "group": ["gmm_group01", "gmm_group02"], "name": "06nonposixgroup", "nonposix": "yes"}, {"description": "test add posix group with members", "name": "06posixgroup", "user": ["gmm_user01", "gmm_user02"]}, {"description": "test add nonposix group with member", "group": ["gmm_group01", "gmm_group02"], "name": "07nonposixgroup", "nonposix": "yes", "user": ["gmm_user01"]}]}, "ansible_included_var_files": ["/root/group_add_multiple_groups.json"], "changed": false}

TASK [Groups] ***********************************************************************************************************************************************************************************************************************
task path: /root/group_module.yml:11
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "group_add: 07nonposixgroup: missing attribute \"gidNumber\" required by object class \"posixGroup\""}
```

Second scenario
```
[root@ansible ~]# cat group_module.yml
---
- name: Playbook to ensure add multiple groups with members using json file.
  hosts: ipaserver
  become: true

  tasks:
  - name: groups add using json file
    include_vars:
      file: group_add_multiple_groups.json  # noqa missing-import

  - name: Groups
    ipagroup:
      ipaadmin_password: <xxxxxxxxxxxxx>
      "groups": "{{ multiplegroups }}"
 
 
[root@ansible ~]# cat group_add_multiple_groups.json
{
  "multiplegroups": [
    {
      "name": "07nonposixgroup",
      "description": "test add nonposix group with member",
      "nonposix": "yes",
      "user": ["gmm_user01"],
      "group": ["gmm_group01", "gmm_group02"]
    }
  ]
}


PLAYBOOK: group_module.yml *******************************************************************************************************************************************************************************
1 plays in group_module.yml

PLAY [Playbook to ensure add multiple groups with members using json file.] ******************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************
task path: /root/group_module.yml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [groups add using json file] ************************************************************************************************************************************************************************
task path: /root/group_module.yml:7
ok: [master.ipadomain.test] => {"ansible_facts": {"multiplegroups": [{"description": "test add nonposix group with member", "group": ["gmm_group01", "gmm_group02"], "name": "07nonposixgroup", "nonposix": "yes", "user": ["gmm_user01"]}]}, "ansible_included_var_files": ["/root/group_add_multiple_groups.json"], "changed": false}

TASK [Groups] ********************************************************************************************************************************************************************************************
task path: /root/group_module.yml:11
changed: [master.ipadomain.test] => {"changed": true}
META: ran handlers
META: ran handlers

PLAY RECAP ***********************************************************************************************************************************************************************************************
master.ipadomain.test      : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0


[root@master ~]# ipa group-show --all 07nonposixgroup
  dn: cn=07nonposixgroup,cn=groups,cn=accounts,dc=ipadomain,dc=test
  Group name: 07nonposixgroup
  Description: test add nonposix group with member
  Member users: gmm_user01
  Member groups: gmm_group02, gmm_group01
  ipauniqueid: efb569d8-b677-11ed-8bf9-fa163e349009
  objectclass: top, groupofnames, nestedgroup, ipausergroup, ipaobject


```

**Comment by dkarpele** [comment]
Thanks @varunmylaraiah , i was able to reproduce the issue. The failing case is:
```
      groups:
      - name: 06posixgroup
      - name: 07nonposixgroup
        nonposix: yes
```
but not vice versa:
```
      groups:
      - name: 07nonposixgroup
        nonposix: yes
      - name: 06posixgroup
```

**Comment by t-woerner** [comment]
I am able to reproduce the issue. Strange is that the issue does not appear using the client context. This really seems to be an IPA issue.

**Comment by rjeffman** [comment]
The performance gain from using batch mode seems to overcome the performance loss from using the client context.

Can we force the client-context for batch mode only? I mean... possible, it is, but do we want it to be done in a way that the default context will be different for batch/non-batch cases?

**Comment by t-woerner** [comment]
Please add this to the README:

Note
-------
Ensuring (adding) several groups with mixed types external, nonposix and posix require to have a fix in IPA. The module implements a work-a-round to automatically switch to the client context if the fix is not applied to IPA and if more than one group is to be ensured with the groups parameter. If `ipaapi_context` is forced to be `server`, the module will fail in this case.

**Comment by dkarpele** [comment]
🙏

File: README-group.md [hunk 5]
```diff
@@ -8,6 +8,9 @@ The group module allows to ensure presence and absence of groups and members of
 
 The group module is as compatible as possible to the Ansible upstream `ipa_group` module, but additionally offers to add users to a group and also to remove users from a group.
 
+## Note
+Ensuring (adding) several groups with mixed types `external`, `nonposix` and `posix` require to have a fix in IPA. The module implements a work-a-round to automatically switch to the `client` context if the fix is not applied to IPA and if more than one group is to be ensured with the groups parameter. If `ipaapi_context` is forced to be `server`, the module will fail in this case.
```
Comment (rjeffman) [inline README-group.md:12]: I suggest a few changes to this text:

> Ensuring presence (adding) of several groups with mixed types (`external`, `nonposix` and `posix`) requires a fix in FreeIPA. The module implements a workaround to automatically use `client` context if the fix is not present in the target node FreeIPA and if more than one group is provided to the task using the `groups` parameter. If `ipaapi_context` is forced to be `server`, the module will fail in this case.

File: README-group.md [hunk 6]
```diff
@@ -71,6 +74,65 @@ Example playbook to add groups:
...
+  "groups": [
+    {
+      "name": "group1",
+      "description": "description group1"
+    },
+    {
+      "name": "group2",
+      "description": "description group2"
+    },
+    ...
```
Comment (rjeffman) [inline README-group.md:111]: The ellipsis is not needed here. I suggest to remove it, along with the previous comma ",".

File: README-group.md [hunk 7]
```diff
@@ -71,6 +74,65 @@ Example playbook to add groups:
       name: appops
 ```
 
+These three `ipagroup` module calls can be combined into one with the `groups` variable:
+
+```yaml
+---
+- name: Playbook to handle groups
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline README-group.md:83]: `become: true` is not needed, please remove it.

File: README-group.md [hunk 8]
```diff
@@ -71,6 +74,65 @@ Example playbook to add groups:
...
+}
+```
+
+And ensure the presence of the groups with this example playbook:
+
+```yaml
+---
+- name: Tests
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline README-group.md:122]: `become: true` is not needed, please remove it.

File: README-group.md [hunk 9]
```diff
@@ -112,11 +174,12 @@ Example playbook to add group members to a group:
 Example playbook to add members from a trusted realm to an external group:
 
 ```yaml
---
+---
 - name: Playbook to handle groups.
   hosts: ipaserver
-  became: true
-
+  become: true
```
Comment (rjeffman) [inline README-group.md:180]: `become: true` is not needed, please remove it.

File: README-group.md [hunk 10]
```diff
@@ -127,6 +190,25 @@ Example playbook to add members from a trusted realm to an external group:
       - WINIPA\\Developers
 ```
 
+Example playbook to add nonposix and external groups:
+
+```yaml
+---
+- name: Playbook to add nonposix and external groups
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline README-group.md:199]: `become: true` is not needed, please remove it.

File: README-group.md [hunk 11]
```diff
@@ -136,13 +218,30 @@ Example playbook to remove groups:
...
       state: absent
 ```
 
+Example playbook to ensure groups are absent:
+
+```yaml
+---
+- name: Playbook to handle groups
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline README-group.md:234]: `become: true` is not needed, please remove it.

File: playbooks/group/add-groups.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+---
+- name: Playbook to handle multiple groups
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline playbooks/group/add-groups.yml:4]: `become: true` is not needed, please remove it.

File: tests/group/test_groups.yml [hunk 13]
```diff
@@ -0,0 +1,144 @@
+---
+- name: Test groups
+  hosts: "{{ ipa_test_host | default('ipaserver') }}"
+  become: yes
```
Comment (rjeffman) [inline tests/group/test_groups.yml:4]: `become: true` is not needed, please remove it.

File: tests/group/test_groups.yml [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  gather_facts: yes
```
Comment (rjeffman) [inline tests/group/test_groups.yml:5]: Please, use only `true` and `false` for boolean values.

File: tests/group/test_groups_absent.yml [hunk 15]
```diff
@@ -0,0 +1,32 @@
+---
+- name: Include create_groups_json.yml
+  ansible.builtin.import_playbook: create_groups_json.yml
+
+- name: Test groups absent
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline tests/group/test_groups_absent.yml:7]: `become: true` is not needed, please remove it.

File: tests/group/test_groups_present.yml [hunk 16]
```diff
@@ -0,0 +1,37 @@
+---
+- name: Include create_groups_json.yml
+  ansible.builtin.import_playbook: create_groups_json.yml
+
+- name: Test groups present
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline tests/group/test_groups_present.yml:7]: `become: true` is not needed, please remove it.

File: tests/group/test_groups_present_slice.yml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+---
+- name: Include create_groups_json.yml
+  ansible.builtin.import_playbook: create_groups_json.yml
+
+- name: Test groups present slice
+  hosts: ipaserver
+  become: true
```
Comment (rjeffman) [inline tests/group/test_groups_present_slice.yml:7]: `become: true` is not needed, please remove it.

**Review by rjeffman — CHANGES_REQUESTED** [review]
Please, on all playbooks (example, documentation and tests), use only `true` and `false` for boolean values.

**Review by rjeffman — APPROVED** [review]
LGTM

File: README-group.md [hunk 18]
```diff
@@ -136,13 +213,29 @@ Example playbook to remove groups:
...
+Example playbook to ensure groups are absent:
+
+```yaml
+---
+- name: Playbook to handle groups
+  hosts: ipaserver
+
+  tanosks:
+  # Ensure groups ops and sysops are absent
+  - ipagroup:
```
Comment (t-woerner) [inline README-group.md:232]: This should be a named task.

File: tests/group/test_groups_absent.yml [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+- name: Test groups absent
+  hosts: ipaserver
+  gather_facts: false
+
+  tasks:
+  - name: Include groups.json
+    ansible.builtin.include_vars:
+      file: groups.json  # noqa 505
+
```
Comment (t-woerner) [inline tests/group/test_groups_absent.yml:13]: Please insert this task:

    - name: Initialize groups_names
      ansible.builtin.set_fact:
        groups_names: []

This is needed if several of the groups tests using `groups_names` are executed in a row.

File: tests/group/test_groups_present.yml [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  tasks:
+  - name: Include groups.json
+    ansible.builtin.include_vars:
+      file: groups.json  # noqa 505
+
+  - name: Groups present len:{{ group_list | length }}
+    ipagroup:
+      ipaadmin_password: SomeADMINpassword
+      groups: "{{ group_list }}"
+
```
Comment (t-woerner) [inline tests/group/test_groups_present.yml:18]: Please insert this task:

    - name: Initialize groups_names
      ansible.builtin.set_fact:
        groups_names: []

This is needed if several of the groups tests using `groups_names` are executed in a row.

File: tests/group/test_groups_present_slice.yml [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+  - name: Size of groups slice.
+    ansible.builtin.debug:
+      msg: "{{ group_list | length }}"
+
+  - name: Groups present
+    ipagroup:
+      ipaadmin_password: SomeADMINpassword
+      groups: "{{ group_list[item : item + slice_size] }}"
+    loop: "{{ range(0, group_list | length, slice_size) | list }}"
+
```
Comment (t-woerner) [inline tests/group/test_groups_present_slice.yml:25]: Please insert this task:

    - name: Initialize groups_names
      ansible.builtin.set_fact:
        groups_names: []

This is needed if several of the groups tests using `groups_names` are executed in a row.

File: plugins/modules/ipagroup.py [hunk 22]
```diff
@@ -43,6 +43,86 @@
     elements: str
     required: true
```
Comment (t-woerner) [inline plugins/modules/ipagroup.py:44]: `name` is not required anymore with this change.

**Comment by t-woerner** [comment]
@dkarpele Please fix the latest findings, then we are able to merge the PR.

**Review by t-woerner — APPROVED** [review]
LGTM

**Pull request merged**